test/extensions/transport_sockets/alts/alts_integration_test.cc

```diff
@@ -0,0 +1,135 @@
+#include "common/common/thread.h"
+
+#include "extensions/transport_sockets/alts/config.h"
+
+#include "test/core/tsi/alts/fake_handshaker/fake_handshaker_server.h"
+#include "test/integration/http_integration.h"
+#include "test/integration/integration.h"
+#include "test/integration/server.h"
+#include "test/integration/utility.h"
+#include "test/mocks/server/mocks.h"
+#include "test/test_common/network_utility.h"
+#include "test/test_common/utility.h"
+
+#include "absl/strings/match.h"
+#include "absl/strings/str_cat.h"
+#include "gmock/gmock.h"
+#include "grpcpp/grpcpp.h"
+#include "grpcpp/impl/codegen/service_type.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Alts {
+
+class AltsIntegrationTest : public HttpIntegrationTest,
+                            public testing::TestWithParam<Network::Address::IpVersion> {
+public:
+  AltsIntegrationTest() : HttpIntegrationTest(Http::CodecClient::Type::HTTP1, GetParam()) {}
+
+  void initialize() override;
+  void SetUp() override;
+  void TearDown() override;
+
+  Network::ClientConnectionPtr makeAltsConnection();
+  std::string fakeHandshakerServerAddress();
+
+  Thread::ThreadPtr fake_handshaker_server_thread_;
+  std::unique_ptr<grpc::Server> fake_handshaker_server_;
+  ConditionalInitializer fake_handshaker_server_ci_;
+  int fake_handshaker_server_port_{};
+  Network::TransportSocketFactoryPtr client_alts_;
+};
+
+std::string AltsIntegrationTest::fakeHandshakerServerAddress() {
+  return absl::StrCat(Network::Test::getLoopbackAddressUrlString(version_), ":",
+                      std::to_string(fake_handshaker_server_port_));
+}
+
+void AltsIntegrationTest::initialize() {
+  config_helper_.addConfigModifier([this](envoy::config::bootstrap::v2::Bootstrap& bootstrap) {
+    auto transport_socket = bootstrap.mutable_static_resources()
+                                ->mutable_listeners(0)
+                                ->mutable_filter_chains(0)
+                                ->mutable_transport_socket();
+    const std::string yaml = R"EOF(
+  name: envoy.transport_sockets.alts
+  config:
+    peer_service_accounts: [peer_identity]
+    handshaker_service: ")EOF" +
+                             fakeHandshakerServerAddress() + "\"";
+
+    MessageUtil::loadFromYaml(TestEnvironment::substitute(yaml), *transport_socket);
+  });
+  HttpIntegrationTest::initialize();
+  registerTestServerPorts({"http"});
+}
+
+void AltsIntegrationTest::SetUp() {
+  fake_handshaker_server_thread_ = std::make_unique<Thread::Thread>([this]() {
+    std::unique_ptr<grpc::Service> service = grpc::gcp::CreateFakeHandshakerService();
+
+    std::string server_address = Network::Test::getLoopbackAddressUrlString(version_) + ":0";
+    grpc::ServerBuilder builder;
+    builder.AddListeningPort(server_address, grpc::InsecureServerCredentials(),
```

Comment:
I'm guessing this is the most expedient way of setting up a fake server. The main downside is that when we hit races and other flakes, Envoy folks are going to have to be become knowledgable about Google gRPC internals, which is a steep cliff.

Reply:
Yes and No, this is almost the only way to setup integration test that can runs simultaneously. (Alternative is run another binary with Python and setup test with env or arg). Given we already have integration test for `GoogleGrpc`, which is subject same races or flakes from Google gRPC and they are dealing with more internal stuff, so I thought this is fine. Thoughts?